### PR TITLE
GPS driver integration

### LINKED ISSUE #54: GPS Module Driver
**Description of feature**

**Requirement related to this feature**

**When is this feature considered "done"**

### PR DESCRIPTION
**Pull request by DanielHernandez-0801**

### CODE REVIEW

File: firmware/src/gps.c [hunk 1]
```diff
@@ -1,18 +1,60 @@
...
     return CLOCK_OK;
 }
-ClockStatus Gps_Connected(Gps *self)
+uint8_t Gps_Connected(Gps *self)
 {
-    UNUSED(self);
-    return CLOCK_OK;
+    gps_connected_flag = (uint8_t) self->gpsConnected();
+
+    curr_time_ms = gpsTimeToMs(Gps_GetTime(self));
```
Comment (Matt-Lemcke) [inline firmware/src/gps.c]: Would it make more sense to check the connection flag before attempting to read the time?

File: firmware/src/gps.c [hunk 2]
```diff
@@ -1,18 +1,60 @@
...
+    return 1;
 }
 GpsTime Gps_GetTime(Gps *self)
 {
-    UNUSED(self);
-    GpsTime zero = { 0 };
-    return zero;
+    // Calculate GPS time from driver UTC time
+    float utc_time = self->getUtcTime();
+    GpsTime t;
```
Comment (Matt-Lemcke) [inline firmware/src/gps.c]: Probably won't be a big deal for us, but we could make t and utc_time global static variables to save on RAM. Up to you if you want to fix this

File: firmware/src/gps.c [hunk 3]
```diff
@@ -1,18 +1,60 @@
...
+    t.hr = (uint8_t) utc_time / 10000;
+    t.min = (uint8_t) ((uint8_t) utc_time % 10000) / 100;
+    t.sec = (uint8_t) utc_time % 100;
+
+    // Apply timezone offset
+    t.hr = t.hr + self->timezone_offset;
```
Comment (Matt-Lemcke) [inline firmware/src/gps.c]: Seems like there could be problems here. Example: t.hr = 23 and offset is 5 would mean t.hr becomes 28. Same if t.hr = 1 and offset is -4 would mean t.hr becomes 252 (unsigned int)

File: firmware/src/gps.c [hunk 4]
```diff
@@ -1,18 +1,60 @@
...
+
+    return t;
+}
+
+ClockStatus Gps_SetTimezone(Gps *self, int8_t timezone)
+{
+    self->timezone_offset = timezone;
+    return CLOCK_OK;
```
Comment (Matt-Lemcke) [inline firmware/src/gps.c]: Maybe add some value checking here and reject values above/below max/min offsets

File: firmware/stm32f4_disc/Core/Inc/uart-gps.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
+/*
+ * gps.h
+ *
+ *  Created on: Nov 15, 2019
+ *      Author: Bulanov Konstantin
+ */
+
+#define	GPS_USART	&huart4
```
Comment (Matt-Lemcke) [inline firmware/stm32f4_disc/Core/Inc/uart-gps.h]: Could we change this so we can choose which uart handler to use? We could do something like i2c-rtc.c where the handler is passed into the Init function

**Review by Matt-Lemcke — CHANGES_REQUESTED** [review]
Just a few comments to look at. Did you update the .ioc file with the changes?

**Comment by DanielHernandez-0801** [comment]
> Just a few comments to look at. Did you update the .ioc file with the changes?

File: firmware/src/gps.c [hunk 6]
```diff
@@ -1,18 +1,74 @@
...
+
+    // Calculate gps hour, min and sec from UTC time and user inputted timezone
+    int8_t hr = ((uint8_t) utc_time / 10000) + self->timezone_offset;
+    if (hr < 0) {
+        gps_time.hr = (uint8_t) 24 + hr;
+    } else if (hr > 23) {
+        gps_time.hr = (uint8_t) hr % 24;
+    } else {
+        gps_time.hr = (uint8_t) hr;
+    }
```
Comment (Matt-Lemcke) [inline firmware/src/gps.c:15]: An equivalent way of writing the hour calculation is gps_time.hr = (24 + utc_hr - offset) % 24. I tried it out and it seems to work well in all the boundary scenarios. Whatever you prefer

File: firmware/src/gps.c [hunk 7]
```diff
@@ -1,18 +1,74 @@
...
 GpsTime Gps_GetTime(Gps *self)
 {
-    UNUSED(self);
-    GpsTime zero = { 0 };
-    return zero;
+    // Get driver UTC time
+    utc_time = self->getUtcTime();
+
+    // Calculate gps hour, min and sec from UTC time and user inputted timezone
+    int8_t hr = ((uint8_t) utc_time / 10000) + self->timezone_offset;
```
Comment (Matt-Lemcke) [inline firmware/src/gps.c]: What is the time unit for utc_time? Based on your calculations it doesn't look like seconds or milliseconds

**Review by Matt-Lemcke — COMMENTED** [review]
Looks good. Just a couple more things before it's ready for testing

File: firmware/src/gps.c (hunk 7 above)
Comment (DanielHernandez-0801) [reply firmware/src/gps.c]: They give the hour, min and seconds as separate digits of the utc time float. For example:

utc_time = 071325.21
hr = 07
min = 13
sec = 25.21

This is why I'm just extracting these digits using integer division and modulo operator. Like so:

utc_time = 071325.21
hr = utc_time / 10000 = 07
min = (utc_time % 10000) / 100 = (1325.21) / 100 = 13
sec = utc_time % 100 = 25.21

Comment (Matt-Lemcke) [reply firmware/src/gps.c]: Ah ok that makes more sense now lol

**Review by Matt-Lemcke — APPROVED** [review]
Ready to merge once we have tested with hardware

**Pull request merged**